File: frontend/src/types/invoice.ts
// frontend/src/types/invoice.ts
export interface InvoiceItem {
  id: string;
  service_type: string;
  date: string;
  start_time: string;
  end_time: string;
  hours: number;
  hourly_rate: number;
  total_amount: number;
  support_worker_name: string;
  notes?: string;
}

export interface Invoice {
  id: string;
  invoice_number: string;
  participant_id: number;
  participant_name: string;
  participant_ndis_number?: string;
  billing_period_start: string;
  billing_period_end: string;
  issue_date: string;
  due_date: string;
  status: 'draft' | 'sent' | 'paid' | 'overdue' | 'cancelled';
  payment_method: 'ndis_direct' | 'plan_managed' | 'self_managed';
  items: InvoiceItem[];
  subtotal: number;
  gst_amount: number;
  total_amount: number;
  amount_paid: number;
  amount_outstanding: number;
  payment_date?: string;
  xero_invoice_id?: string;
  created_at: string;
  updated_at?: string;
}

export interface Payment {
  id: string;
  invoice_id: string;
  amount: number;
  payment_date: string;
  payment_method: string;
  reference_number?: string;
  notes?: string;
  processed_by?: string;
  xero_payment_id?: string;
  created_at?: string;
}

export interface InvoiceStats {
  total_invoices: number;
  total_outstanding: number;
  total_overdue: number;
  total_paid_this_month: number;
  average_payment_days: number;
}

export interface BillableService {
  id: string;
  appointment_id: number;
  participant_id: number;
  participant_name: string;
  service_type: string;
  date: string;
  start_time: string;
  end_time: string;
  hours: number;
  hourly_rate: number;
  total_amount: number;
  support_worker_name: string;
  notes?: string;
  is_billable: boolean;
  invoice_id?: string;
  created_at: string;
}

export interface InvoiceGenerationRequest {
  participant_id?: number;
  billing_period_start: string;
  billing_period_end: string;
  issue_date: string;
  due_date: string;
  payment_method: string;
  service_ids: string[];
  include_gst: boolean;
  gst_rate: number;
  auto_send: boolean;
  notes?: string;
}

export interface XeroSyncStatus {
  connected: boolean;
  last_sync?: string;
  tenant_name?: string;
  sync_status: 'idle' | 'syncing' | 'error' | 'success';
  error_message?: string;
}

export interface PaymentStats {
  total_received_this_month: number;
  total_outstanding: number;
  average_payment_days: number;
  overdue_amount: number;
  pending_payments: number;
}

// For payment tracking with invoice details
export interface PaymentWithInvoice extends Payment {
  invoice_number: string;
  participant_name: string;
  participant_id: number;
  invoice_total: number;
}

File: backend/app/models/invoice.py
# backend/app/models/invoice.py
from sqlalchemy import Column, Integer, String, Text, Date, DateTime, Numeric, ForeignKey, Enum
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import enum
from app.core.database import Base

class InvoiceStatus(str, enum.Enum):
    draft = "draft"
    sent = "sent"
    paid = "paid"
    overdue = "overdue"
    cancelled = "cancelled"

class PaymentMethod(str, enum.Enum):
    ndis_direct = "ndis_direct"
    plan_managed = "plan_managed"
    self_managed = "self_managed"

class Invoice(Base):
    __tablename__ = "invoices"

    id = Column(Integer, primary_key=True, index=True)
    invoice_number = Column(String(50), unique=True, nullable=False, index=True)

    # Participant info
    participant_id = Column(Integer, ForeignKey("participants.id"), nullable=False)
    participant_name = Column(String(200), nullable=False)
    participant_ndis_number = Column(String(50), nullable=True)

    # Billing period
    billing_period_start = Column(Date, nullable=False)
    billing_period_end = Column(Date, nullable=False)

    # Invoice dates
    issue_date = Column(Date, nullable=False)
    due_date = Column(Date, nullable=True)

    # Status and payment
    status = Column(Enum(InvoiceStatus), nullable=False, default=InvoiceStatus.draft)
    payment_method = Column(Enum(PaymentMethod), nullable=False, default=PaymentMethod.ndis_direct)

    # Amounts
    subtotal = Column(Numeric(10, 2), nullable=False)
    gst_amount = Column(Numeric(10, 2), nullable=False, default=0)
    total_amount = Column(Numeric(10, 2), nullable=False)
    amount_paid = Column(Numeric(10, 2), nullable=False, default=0)
    amount_outstanding = Column(Numeric(10, 2), nullable=False)

    # Payment tracking
    payment_date = Column(Date, nullable=True)

    # External references
    xero_invoice_id = Column(String(100), nullable=True)

    # Additional info
    notes = Column(Text, nullable=True)

    # Audit
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, onupdate=func.now())
    created_by = Column(Integer, ForeignKey("users.id"), nullable=True)

    # Relationships
    items = relationship("InvoiceItem", back_populates="invoice", cascade="all, delete-orphan")
    participant = relationship("Participant")

class InvoiceItem(Base):
    __tablename__ = "invoice_items"

    id = Column(Integer, primary_key=True, index=True)
    invoice_id = Column(Integer, ForeignKey("invoices.id"), nullable=False)

    # Link to roster/appointment
    appointment_id = Column(Integer, nullable=True)

    # Service details
    service_type = Column(String(100), nullable=False)
    date = Column(Date, nullable=False)
    start_time = Column(String(10), nullable=False)
    end_time = Column(String(10), nullable=False)
    hours = Column(Numeric(10, 2), nullable=False)
    hourly_rate = Column(Numeric(10, 2), nullable=False)
    total_amount = Column(Numeric(10, 2), nullable=False)

    # Support worker
    support_worker_name = Column(String(200), nullable=True)

    # Notes
    notes = Column(Text, nullable=True)

    # Relationship
    invoice = relationship("Invoice", back_populates="items")
